sklearn/metrics/_plot/confusion_matrix.py

```diff
@@ -175,20 +179,30 @@ def plot_confusion_matrix(estimator, X, y_true, labels=None,
         Axes object to plot on. If `None`, a new figure and axes is
         created.
 
+    y_pred : array-like of shape (n_samples,), default=None
+        Predicted labels. Either `estimator` and `X` must be specified, or
+        `y_pred`.
+
     Returns
     -------
     display : :class:`~sklearn.metrics.ConfusionMatrixDisplay`
     """
     check_matplotlib_support("plot_confusion_matrix")
 
-    if not is_classifier(estimator):
-        raise ValueError("plot_confusion_matrix only supports classifiers")
+    if estimator is not None and X is not None and y_pred is None:
+        if not is_classifier(estimator):
+            raise ValueError("plot_confusion_matrix only supports classifiers")
+        y_pred = estimator.predict(X)
+    elif estimator is None and X is None and y_pred is not None:
+        check_consistent_length(y_true, y_pred)
+    else:
+        raise ValueError("Either 'estimator' and 'X' must be passed to "
+                         "plot_confusion_matrix or 'y_pred'")
```

Comment:
Please add tests for:

- check the error message
- check that passing y_pred gives the same confusion matrix as passing the estimator and X that generated the predictions
- check that a proper error message is raised if y_true is left to None